kms/awskms/decrypter.go

```diff
@@ -0,0 +1,172 @@
+func NewDecrypter(client KeyManagementClient, decryptionKey string) (*Decrypter, error) {
+	keyID, err := parseKeyID(decryptionKey)
+	if err != nil {
+		return nil, err
+	}
+
+	decrypter := &Decrypter{
+		client: client,
+		keyID:  keyID,
+	}
+	if err := decrypter.preloadKey(); err != nil {
+		return nil, err
+	}
+
+	return decrypter, nil
+}
+
+func (d *Decrypter) preloadKey() error {
+	ctx, cancel := defaultContext()
+	defer cancel()
+
+	resp, err := d.client.GetPublicKey(ctx, &kms.GetPublicKeyInput{
+		KeyId: pointer(d.keyID),
+	})
+	if err != nil {
+		return fmt.Errorf("awskms GetPublicKey failed: %w", err)
+	}
+
+	d.publicKey, err = pemutil.ParseDER(resp.PublicKey)
+	return err
+}
+
+// Public returns the public key of this decrypter
+func (d *Decrypter) Public() crypto.PublicKey {
+	return d.publicKey
+}
+
+// Decrypt decrypts ciphertext using the decryption key backed by AWS KMS and returns
+// the plaintext bytes. An error is returned when decryption fails. AWS KMS only supports
+// RSA keys with 2048, 3072 or 4096 bits and will always use OAEP. It supports SHA1 and SHA256.
+// Labels are not supported. Before calling out to AWS, some validation is performed
+// so that known bad parameters are detected client-side and a more meaningful error is returned
+// for those cases.
+//
+// Also see https://docs.aws.amazon.com/kms/latest/developerguide/symm-asymm-choose-key-spec.html#key-spec-rsa.
+func (d *Decrypter) Decrypt(_ io.Reader, ciphertext []byte, opts crypto.DecrypterOpts) ([]byte, error) {
+	algorithm, err := determineDecryptionAlgorithm(d.publicKey, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed determining decryption algorithm: %w", err)
+	}
+
+	req := &kms.DecryptInput{
+		KeyId:               pointer(d.keyID),
+		CiphertextBlob:      ciphertext,
+		EncryptionAlgorithm: algorithm,
+	}
+
+	ctx, cancel := defaultContext()
+	defer cancel()
+
+	response, err := d.client.Decrypt(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("awskms Decrypt failed: %w", err)
+	}
+
+	// TODO: additional validation?
+
+	return response.Plaintext, nil
+}
+
+const (
+	awsOaepSha1   = "RSAES_OAEP_SHA_1"
+	awsOaepSha256 = "RSAES_OAEP_SHA_256"
+)
+
+func determineDecryptionAlgorithm(key crypto.PublicKey, opts crypto.DecrypterOpts) (types.EncryptionAlgorithmSpec, error) {
+	pub, ok := key.(*rsa.PublicKey)
+	if !ok {
+		return "", fmt.Errorf("awskms does not support key type %T", key)
+	}
+
+	if opts == nil {
+		opts = &rsa.OAEPOptions{}
+	}
+
+	var rsaOpts *rsa.OAEPOptions
+	switch o := opts.(type) {
+	case *rsa.OAEPOptions:
+		if err := validateOAEPOptions(o); err != nil {
+			return "", err
+		}
+		rsaOpts = o
+	case *rsa.PKCS1v15DecryptOptions:
+		return "", errors.New("awskms does not support PKCS #1 v1.5 decryption")
+	default:
+		return "", fmt.Errorf("invalid decrypter options type %T", opts)
+	}
+
+	switch pub.Size() {
+	default:
+		return "", fmt.Errorf("awskms does not support RSA public key size %d", pub.Size())
```

Comment:
```suggestion

        bitSize := pub.Size() * 8
	switch bitSize {
	default:
		return "", fmt.Errorf("awskms does not support RSA public key size %d", bitSize)
```
`pub.Size()` is in bytes so the switch will always say we don't have the right key size.

Reply:
Thanks! 7c54eedf31a33db2627261138843e90df7fc1c82 🙂